### ISSUE TITLE
Nashorn object['func'] cannot be invoked with call()

### ISSUE THREAD

**Issue by sbordet** [cometd/cometd#742]
@szedegi referring to your comment in #740, the reason `CometDDojoHitchTest.testDojoHitch()` fails is the following:

```javascript
var latch = new Latch(1);
var countDown = latch['countDown'];
print(countDown, typeof countDown);
print('call', countDown.call, typeof countDown.call);
print('apply', countDown.apply, typeof countDown.apply);
countDown.call();
```

which prints:

```
[jdk.internal.dynalink.beans.SimpleDynamicMethod void org.cometd.javascript.Latch.countDown()] function
call undefined undefined
apply undefined undefined
```

and of course fails trying to execute `call()` or `apply()`.

So, when extracting a method from an object I get back something whose `typeof` says it's a `function`, but then I cannot invoke `call()` or `apply()` on it.

Nashorn bug ?

**Comment by szegedi** [comment]
I remember there being a reason we didn't expose `call` and `apply` on non-native function (meaning, they're really host objects that happen to be callables working with `obj.method()` syntax, but they don't inherit `Function.prototype`…

A workaround you could use is: `Function.prototype.call.call(countDown, latch)`. Note that even if `.call` worked directly, you'd still need to pass it a this - `countDown` is not bound to `latch`. I honestly don't remember whether we implemented `bind` for arbitrary callables, but if we did then you could also try:

```
var countDown = Function.prototype.bind.call(latch['countDown'], latch)
```
and then invoke `countDown()` as a bound function (`Function.apply` and `.call` in Nashorn have been enhanced to handle any callables, not just native functions; I don't remember off the top of my head whether we did the same with `.bind`)

**Comment by szegedi** [comment]
Oh, I remember why we don't expose `.call` and `.apply` on every Nashorn callable… Basically, Nashorn treats a bunch of objects as being "callable", meaning it reacts to the `()` operator – JS functions, obviously, but also methods of Java objects, Java objects implementing an interface with the `@FunctionalInterface` annotation, and so on.

And herein lies the rub: `j.u.function.Function` already has an `apply` method; `j.u.c.Callable` already has a `call` method. If we started magically exposing JS-semantics `.call` and `.apply` on them, we'd be clobbering their own same-named methods. We decided to resolve the conflict by not pretending all callables inherit `Function` prototype methods, but we made them work if invoked explicitly on them as per above (`Function.prototype.call.call` etc.)
